src/module-config/module-config.ts

```diff
@@ -6,6 +6,17 @@ const configs: object[] = [];
 // An object with module names for keys and schemas for values.
 const schemas = {};
 
+// Pull default config file from module-config
+const configPromise = System.import("module-config");                
```

Comment:
Jest complains:

```
FAIL  src/module-config/module-config.test.ts
 ● Test suite failed to run

   TypeError: System.import is not a function
```

Despite the `setup-tests.js` and `jest.config.js` seen here.

Reply:
So I'm a bit confused here -- we're assuming that there is a module called `module-config` that is separate from `@openmrs/esm-module-config`? But we also have a `provide()` function for adding more configs?

I was under the impression that people would call `provide()` from inside of their root config with whatever config they wanted to load, instead of assuming a module name.

Regarding the test failure, it looks like you've mocked System.resolve and System.register but haven't mocked `System.import()`, which is the one that is being used.

Reply:
> Regarding the test failure, it looks like you've mocked System.resolve and System.register but haven't mocked System.import(), which is the one that is being used.

D'oh, thanks!

Regarding `provide()` vs the imported config: the idea is that people shouldn't have to override the root-config in order to configure. They should be able to use MF out of the box, and then by modifying the import map, load their own config file. My goal here is to lower the (technical skill) barrier to entry for using MF.

If an implementer wants to do anything fancier, e.g. with hierarchical or dynamic or whatever configs, they will need a root-config override and to use the `provide` API.

The import-mapped config file is not mandatory (I would like eventually to squelch the 404 error on the console). If used in conjunction with `provide` it will simply be the lowest-priority config file.

I have no idea what to name the import mapped config file -- so far "openmrs-config" and "module-config" have been candidates. Recommendations very welcome.

Reply:
The names are going to confuse people imo

@openmrs/esm-module-config
@openmrs/esm-root-config
module-config

^ knowing the difference between these three will be hard.

Do you think it’ll be possible for people to realistically not have their own root config?

I think that it might be okay to assume a name of module to use as config, but that it should be named something distinct enough for people to be able to understand it immediately without having to look up the difference between esm-module-config and module-config. I also think that we should not require that module to be present in the import map at all and still have everything work. In other words, it’s a default that can be used but it doesn’t have to be used.

Maybe a name that would be easier would be custom-implementation-config? Or something?

Reply:
Btw I’m pretty sure you can check if a module is in the import map with System.resolve, which I think throws an error of the provided name is not in the import map. See https://github.com/systemjs/systemjs/blob/master/docs/api.md#systemresolveid--parenturl---string for more details.

Also “custom-module-config” could be another option for naming

Reply:
how about @openmrs/esm-distro-config?

Reply:
It won't be under the `@openmrs` npm organization, and also doesn't need to be an ES module (can be a JSON module).

Reply:
Oh, how about just `config-file`?

> Btw I’m pretty sure you can check if a module is in the import map with System.resolve, 

Beautiful, thank you!

> Do you think it’ll be possible for people to realistically not have their own root config?

I have no idea. I hope so? That would seem like a requirement for keeping with the promise that OpenMRS should be usable without a software engineer on hand.

> @openmrs/esm-module-config
> @openmrs/esm-root-config

I wonder if we should have called this `esm-[module-]config-lib[rary]`.

Reply:
I agree with rename of module to config-lib.

Reply:
@joeldenning , whaddya think?

Reply:
`@openmrs/esm-module-lib` and `config-file`?

Reply:
👍 config file sounds good to me